js/src/dom/data.js

```diff
@@ -11,57 +11,45 @@
  * ------------------------------------------------------------------------
  */
 
-const mapData = (() => {
-  const storeData = {}
-  let id = 1
-  return {
-    set(element, key, data) {
-      if (typeof element.bsKey === 'undefined') {
-        element.bsKey = {
-          key,
-          id
-        }
-        id++
-      }
+const elementMap = new WeakMap()
 
-      storeData[element.bsKey.id] = data
-    },
-    get(element, key) {
-      if (!element || typeof element.bsKey === 'undefined') {
-        return null
-      }
+export default {
+  set(element, key, instance) {
+    if (!elementMap.has(element)) {
+      elementMap.set(element, new Map())
+    }
 
-      const keyProperties = element.bsKey
-      if (keyProperties.key === key) {
-        return storeData[keyProperties.id]
-      }
+    const instanceMap = elementMap.get(element)
 
-      return null
-    },
-    delete(element, key) {
-      if (typeof element.bsKey === 'undefined') {
-        return
-      }
-
-      const keyProperties = element.bsKey
-      if (keyProperties.key === key) {
-        delete storeData[keyProperties.id]
-        delete element.bsKey
-      }
+    // make it clear we only want one instance per element
+    // can be removed at a later time when multiple key/instances are allowed
+    if (!instanceMap.has(key) && instanceMap.size !== 0) {
+      throw new Error(
+        'Sorry, we currently allow only one instance per element.'
+      )
```

Comment:
It would be better if you use bootstrap here instead of we

Reply:
Not sure if we should throw an error. As much as I like this approach, we generally try to not break people's code.

Reply:
@XhmikosR in general I'm with you on this, but I'm not sure what's the best way to deal with this.
Throwing an error has the advantage of making tests fail in case someone was to implement/refactor a plugin while using multiple instances on an element without being aware of it.
From another perspective the old implementation had kind of a bug, as it wouldn't update the `key` properly when used with a different `key` on any additional call to `setData`. That likely to cause some implications down the line, even when used in production code (maybe without even noticing). So an error could be helpful to detect there is something wrong here.
As an alternative to an error, what do you think would be a good way? We could use console.error and return without altering the map, or do we want to show just the information and continue? We could also mimic the old behaviour, but that means we reintroduce that `key` flaw again.

Reply:
Just so that we are on the same boat, I don't believe that no feedback is a good approach. That being said, if we are going to throw an error, this change needs to be in beta1 which is soon to be released, after documenting it in migration notes.

Otherwise, we can't land it later, only in a major version bump.